### PR TITLE
Created New Template for Deploying VMSS with Public IP Prefix

### PR DESCRIPTION
**Pull request by allegradomel**
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use a parameter for resource locations with the defaultValue set to resourceGroup().location
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values)
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

*
*
*

### CODE REVIEW

File: 101-vmms-with-public-ip-prefix/README.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
+# VMSS with Public IP Prefix 
+
+<IMG SRC="https://azurequickstartsservice.blob.core.windows.net/badges/100-blank-template/PublicLastTestDate.svg" />&nbsp;
```
Comment (bmoore-msft) [inline 101-vmms-with-public-ip-prefix/README.md]: update these with the path to your sample

File: 101-vmms-with-public-ip-prefix/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+      }
+    },
+    "adminPassword": {
+      "type": "securestring",
+      "metadata": {
+        "description": "Admin password on all VMs."
+      }
+    },
+    "publicIPPrefixLength": {
+      "type": "string",
```
Comment (bmoore-msft) [inline 101-vmms-with-public-ip-prefix/azuredeploy.json]: should this be an int? with min/max values?

File: 101-vmms-with-public-ip-prefix/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+      "metadata":{
+        "description":"Length of public IP prefix."
+      }
+    }
+    
+  },
+  "variables": {
+    "location": "[resourceGroup().location]",
```
Comment (bmoore-msft) [inline 101-vmms-with-public-ip-prefix/azuredeploy.json]: Add location parameter - see: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#parameters

File: 101-vmms-with-public-ip-prefix/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+    "publicIPAddressID": "[resourceId('Microsoft.Network/publicIPAddresses',variables('publicIPAddressName'))]",
+    "lbID": "[resourceId('Microsoft.Network/loadBalancers',variables('loadBalancerName'))]",
+    "natPoolName": "[concat(parameters('vmssName'), 'natpool')]",
+    "bePoolName": "[concat(parameters('vmssName'), 'bepool')]",
+    "natStartPort": 50000,
+    "natEndPort": 50120,
+    "natBackendPort": 22,
+    "nicName": "[concat(parameters('vmssName'), 'nic')]",
+    "ipConfigName": "[concat(parameters('vmssName'), 'ipconfig')]",
+    "frontEndIPConfigID": "[concat(variables('lbID'),'/frontendIPConfigurations/loadBalancerFrontEnd')]",
```
Comment (bmoore-msft) [inline 101-vmms-with-public-ip-prefix/azuredeploy.json]: remove concat and use resourceId

File: 101-vmms-with-public-ip-prefix/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+    },
+    "imageReference": "[variables('osType')]"
+  },
+  "resources": [
+    {
+      "type": "Microsoft.Network/publicIPPrefixes",
+      "apiVersion": "2018-07-01",
+      "name": "[variables('publicIPPrefixName')]",
+      "location": "[variables('location')]",
+      "dependsOn": [],
```
Comment (bmoore-msft) [inline 101-vmms-with-public-ip-prefix/azuredeploy.json]: remove all empty/null properties

File: 101-vmms-with-public-ip-prefix/azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+        ]
+      }
+    },
+    {
+      "type": "Microsoft.Network/publicIPAddresses",
+      "name": "[variables('publicIPAddressName')]",
+      "location": "[variables('location')]",
+      "apiVersion": "2018-08-01",
+        "sku": {
+    "name": "Standard",
```
Comment (bmoore-msft) [inline 101-vmms-with-public-ip-prefix/azuredeploy.json]: format the json (e.g. SHIFT+ALT+F in VS Code)

File: 101-vmms-with-public-ip-prefix/azuredeploy.json [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+              {
+                "name": "[variables('nicName')]",
+                "properties": {
+                  "primary": true,
+                  "ipConfigurations": [
+                    {
+                      "name": "[variables('ipConfigName')]",
+                      "properties": {
+                        "subnet": {
+                          "id": "[concat('/subscriptions/', subscription().subscriptionId,'/resourceGroups/', resourceGroup().name, '/providers/Microsoft.Network/virtualNetworks/', variables('virtualNetworkName'), '/subnets/', variables('subnetName'))]"
```
Comment (bmoore-msft) [inline 101-vmms-with-public-ip-prefix/azuredeploy.json]: remove concat and use resourceId

File: 101-vmms-with-public-ip-prefix/azuredeploy.json [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+                        },
+                        "publicipaddressconfiguration": {
+                          "name": "pub1",
+                          "properties": {
+                            "idleTimeoutInMinutes": 15,
+                            "PublicIpPrefix":  {
+                                "Id": "[concat('/subscriptions/', subscription().subscriptionId,'/resourceGroups/', resourceGroup().name, '/providers/Microsoft.Network/publicIPPrefixes/', variables('publicIpPrefixName'))]"
```
Comment (bmoore-msft) [inline 101-vmms-with-public-ip-prefix/azuredeploy.json]: remove concat and use resourceId

File: 101-vmms-with-public-ip-prefix/azuredeploy.json [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+                            }
+                          }
+                        },
+                        "loadBalancerBackendAddressPools": [
+                          {
+                            "id": "[concat('/subscriptions/', subscription().subscriptionId,'/resourceGroups/', resourceGroup().name, '/providers/Microsoft.Network/loadBalancers/', variables('loadBalancerName'), '/backendAddressPools/', variables('bePoolName'))]"
```
Comment (bmoore-msft) [inline 101-vmms-with-public-ip-prefix/azuredeploy.json]: remove concat and use resourceId

File: 101-vmms-with-public-ip-prefix/azuredeploy.json [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+                          }
+                        ],
+                       "loadBalancerInboundNatPools": [
+                          {
+                            "id": "[concat('/subscriptions/', subscription().subscriptionId,'/resourceGroups/', resourceGroup().name, '/providers/Microsoft.Network/loadBalancers/', variables('loadBalancerName'), '/inboundNatPools/', variables('natPoolName'))]"
```
Comment (bmoore-msft) [inline 101-vmms-with-public-ip-prefix/azuredeploy.json]: remove concat and use resourceId

File: 101-vmms-with-public-ip-prefix/azuredeploy.parameters.json [hunk 11]
```diff
@@ -0,0 +1,24 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentParameters.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
```
Comment (bmoore-msft) [inline 101-vmms-with-public-ip-prefix/azuredeploy.parameters.json]: remove the non-GEN parameters and use defaultValues in the template

File: 101-vmms-with-public-ip-prefix/metadata.json [hunk 12]
```diff
@@ -0,0 +1,9 @@
+{
+  "$schema": "https://schema.management.azure.com/schemas/2015-01-01/...",
```
Comment (bmoore-msft) [inline 101-vmms-with-public-ip-prefix/metadata.json]: this is not a valid $schema property (see other samples)

Comment (allegradomel) [reply 101-vmms-with-public-ip-prefix/metadata.json]: Changed to valid $schema property as found in other templates

**Comment by allegradomel** [comment]
@bmoore-msft, I have made the above changes

**Pull request merged**